File: crates/sandbox/src/bin/weather/sky/sun.f90
! 2019-12-26 ! Solar Geometry using subsolar point and atan2.
!              by Taiping Zhang.
!
! Input variables:
!     inyear: 4-digit year, e.g., 1998, 2020;
!     inmon: month, in the range of 1 - 12;
!     inday: day, in the range 1 - 28/29/30/31;
!     gmtime: GMT in decimal hour, e.g., 15.2167;
!     xlat: latitude in decimal degree, positive in Northern Hemisphere;
!     xlon: longitude in decimal degree, positive for East longitude.
!
! Output variables:
!     solarz: solar zenith angle in deg;
!     azi: solar azimuth in deg the range -180 to 180, South-Clockwise
!          Convention.
!
! Note: The user may modify the code to output other variables.

Subroutine sunpos_ultimate_azi_atan2(inyear, inmon, inday, gmtime, xlat, xlon, solarz, azi)
   implicit none

   integer:: inyear, inmon, inday, nday(12), julday(0:12), xleap, i, dyear, dayofyr
   real:: gmtime, xlat, xlon
   real:: n, L, g, lambda, epsilon, alpha, delta, R, EoT
   real:: solarz, azi
   real, parameter:: rpd = acos(-1.0)/180
   real:: sunlat, sunlon, PHIo, PHIs, LAMo, LAMs, Sx, Sy, Sz
   data nday/31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31/

   if ((mod(inyear, 100) /= 0 .and. mod(inyear, 4) == 0) .or. &
       (mod(inyear, 100) == 0 .and. mod(inyear, 400) == 0)) then
      nday(2) = 29
   else
      nday(2) = 28
   end if

   julday(0) = 0
   do i = 1, 12
      julday(i) = julday(i - 1) + nday(i)
   end do
   ! Note: julday(12) is equal to either 365 or 366.

   dyear = inyear - 2000
   dayofyr = julday(inmon - 1) + inday
   if (dyear <= 0) then
      xleap = int(real(dyear)/4) ! Note: xleap has the SAME SIGN as dyear. !!!
   elseif (dyear > 0) then
      if (mod(dyear, 4) == 0) then
         xleap = int(real(dyear)/4) !  For leap-years.
      else
         xleap = int(real(dyear)/4) + 1 ! "+1" is for year 2000.
      end if
   end if

   ! --- Astronomical Almanac for the Year 2019, Page C5 ---
   n = -1.5 + dyear*365.0 + xleap*1.0 + dayofyr + gmtime/24
   L = modulo(280.460 + 0.9856474*n, 360.0)
   g = modulo(357.528 + 0.9856003*n, 360.0)
   lambda = modulo(L + 1.915*sin(g*rpd) + 0.020*sin(2*g*rpd), 360.0)
   epsilon = 23.439 - 0.0000004*n
   alpha = modulo(atan2(cos(epsilon*rpd)*sin(lambda*rpd), &
                        cos(lambda*rpd))/rpd, 360.0) ! alpha in the same quadrant as lambda.
   delta = asin(sin(epsilon*rpd)*sin(lambda*rpd))/rpd
   R = 1.00014 - 0.01671*cos(g*rpd) - 0.00014*cos(2*g*rpd)
   EoT = modulo((L - alpha) + 180.0, 360.0) - 180.0 ! In deg.

   ! --- Solar geometry ---
   sunlat = delta ! In deg.
   sunlon = -15.0*(gmtime - 12.0 + EoT*4/60)
   PHIo = xlat*rpd
   PHIs = sunlat*rpd
   LAMo = xlon*rpd
   LAMs = sunlon*rpd
   Sx = cos(PHIs)*sin(LAMs - LAMo)
   Sy = cos(PHIo)*sin(PHIs) - sin(PHIo)*cos(PHIs)*cos(LAMs - LAMo)
   Sz = sin(PHIo)*sin(PHIs) + cos(PHIo)*cos(PHIs)*cos(LAMs - LAMo)

   solarz = acos(Sz)/rpd ! In deg.
   azi = atan2(-Sx, -Sy)/rpd ! In deg. South-Clockwise Convention.
End subroutine sunpos_ultimate_azi_atan2

File: crates/sandbox/src/bin/weather/sky/sun.rs
#![allow(non_snake_case)]

use chrono::prelude::*;

// https://en.wikipedia.org/wiki/Solar_zenith_angle
// https://en.wikipedia.org/wiki/Solar_azimuth_angle
// https://doi.org/10.1016/j.renene.2021.03.047
pub fn zenith_azimuth(utc: NaiveDateTime, lat: f64, lon: f64) -> (f64, f64) {
  // Astronomical Almanac for the Year 2019, Page C5
  // https://archive.org/details/binder1_202003/page/n213/mode/1up
  let n = days_since_J2000(utc);
  let L = (280.460 + 0.9856474 * n).to_radians();
  let g = (357.528 + 0.9856003 * n).to_radians();
  let λ = L + (1.915 * g.sin() + 0.020 * (2.0 * g).sin()).to_radians();
  let ϵ = (23.439 - 0.0000004 * n).to_radians();
  let α = (ϵ.cos() * λ.sin()).atan2(λ.cos());
  let δ = (ϵ.sin() * λ.sin()).asin();
  let E = L - α;

  // the equation of time in minutes and the GMT in hours
  let E_min = (E.to_degrees() + 180.0) % 360.0 - 180.0;
  let T_GMT = hours_since_midnight(utc);

  // the latitude and longitude of the observer
  let ϕo = lat.to_radians();
  let λo = lon.to_radians();

  // the latitude and longitude of the subsolar point
  let ϕs = δ;
  let λs = -15.0 * (T_GMT - 12.0 + E_min / 60.0).to_radians();

  // the unit vector pointing toward the Sun
  let Sx = ϕs.cos() * (λs - λo).sin();
  let Sy = ϕo.cos() * ϕs.sin() - ϕo.sin() * ϕs.cos() * (λs - λo).cos();
  let Sz = ϕo.sin() * ϕs.sin() + ϕo.cos() * ϕs.cos() * (λs - λo).cos();

  // the solar zenith angle
  let Z = Sz.acos();
  let Z = Z.to_degrees();

  // the solar azimuth angle
  // let γs = Sy.atan2(Sx); // East-Counterclockwise Convention
  // let γs = Sx.atan2(Sy); // North-Clockwise Convention
  let γs = (-Sx).atan2(-Sy); // South-Clockwise Convention
  let γs = γs.to_degrees();

  (Z, γs)
}

fn days_since_J2000(dt: NaiveDateTime) -> f64 {
  let d = NaiveDate::from_yo_opt(2000, 1).unwrap();
  let t = NaiveTime::from_hms_opt(12, 0, 0).unwrap();
  let dur = dt.signed_duration_since(d.and_time(t));
  dur.num_seconds() as f64 / (24 * 60 * 60) as f64
}

fn hours_since_midnight(dt: NaiveDateTime) -> f64 {
  dt.num_seconds_from_midnight() as f64 / (60 * 60) as f64
}
